### ISSUE TITLE
`isConjugated` is `false` for Arginine terminal Ns in PDBs

### ISSUE THREAD

**Issue by pechersky** [molstar/molstar#1361]
https://github.com/molstar/molstar/blob/c53b651472c63adaf1115753375bdea6e5b56045/src/mol-model-props/computed/chemistry/valence-model.ts#L40-L67

This affects H-bond calculation, because the ideal donor angle is then based off of tetrahedral (109) instead of the correct trigonal (120) geometry. The `isConjugated` calculation gives `false` because in PDBs, there are no bond orders, so there are no bonds with `bond.order > 1`.

**Comment by arose** [comment]
We assign bond order for known cases, see src\mol-model\structure\model\properties\atomic\bonds.ts

There is a comment to interpret connect records (if given in a non-standard way) in PDB files to assign bond orders, see src\mol-model-formats\structure\pdb\conect.ts

Maybe we can also do a distance check in `isConjugated`? Something base on section 9 in https://www.daylight.com/meetings/mug01/Sayle/m4xbondage.html

> unsatisfied sp2-hybridized atom by checking each neighboring atom for a terminal oxygen, and when found testing the bond length against the table of bond lengths

Feel free to do a PR!

**Comment by pechersky** [comment]
Not sure what the best way to overcome this would be. Somehow in the 3D viewer, the Arg does show a double bond to one of the terminal Ns in the arginine -- so somewhere bond orders are inferred. Is it possible to run the `isConjugated` after that?

Additionally, in our case, we would satisfy the `donHAngle` check -- given that the Hs exist, why should we check the `X -- N` angle at all? Maybe we only check `donAngle` within ideal angle if Hs don't exist?

Short term, we're going to hardcode `NH1` and `NH2` to return `isConjugated` true just so that they get trigonal geom.

**Comment by papillot** [comment]
> Additionally, in our case, we would satisfy the `donHAngle` check -- given that the Hs exist, why should we check the `X -- N` angle at all? Maybe we only check `donAngle` within ideal angle if Hs don't exist?
>

There are cases where the H atom position is not compatible with a putative H-bond. It is necessary to combine both data (i.e. "Are there donors and acceptors at suitable positions?", "When Hydrogens are present, do they contradict the H-bond perception?" ) when they are present to discard false positive.

**Comment by pechersky** [comment]
In my case, I have "Hs are present, they should be the only determinants of angles". I read your statement as:
```
don and acceptors look okay
oh wait, Hs aren't placed in a way that is acceptable
```
so the latter cancels the former. While my issue here is
```
don and acceptors don't look okay, due to wrong ideal angle
never even hit the H-based check.
```

**Comment by arose** [comment]
Can you share an example file/residue? The bond order assignment depends on specific atoms names (as used in PDB archive and various xtal software). If you deviate from that and assign an H to an arginine with a double bond things probably go wrong.

**Comment by pechersky** [comment]
Here is an extracted ARG and ligand portion. In the original PDB file, there were no CONECT for any of the intra-ARG bonds.
[1361_example.zip](https://github.com/user-attachments/files/17908350/1361_example.zip)

**Comment by pechersky** [comment]
For further clarification, OpenEye tools mark that as a `hbond:protein2ligand` interaction.

**Comment by pechersky** [comment]
```
$ cat perceive_hints.py
import openeye.oechem as oe

def GetInteractionString(inter):

    fragstrs = []
    for frag in [inter.GetBgnFragment(), inter.GetEndFragment()]:
        if frag is None:
            continue
        fragtype = frag.GetComponentType()
        if fragtype == oe.OELigandInteractionHintComponent():
            fragstrs.append("ligand:" + " ".join(sorted(str(a) for a in frag.GetAtoms())))
        if fragtype == oe.OEProteinInteractionHintComponent():
            fragstrs.append("protein:" + " ".join(sorted(str(a) for a in frag.GetAtoms())))

    return " ".join(sorted(f for f in fragstrs))
ifs = oe.oemolistream("1361_example.pdb")
protein = next(iter(ifs.GetOEGraphMols()))
sifs = oe.oemolistream("1361_example.sdf")
ligand = next(iter(sifs.GetOEGraphMols()))
asite = oe.OEInteractionHintContainer(protein, ligand)
oe.OEPerceiveInteractionHints(asite)
for interaction in asite.GetInteractions():
    print(interaction.GetInteractionType().GetName())
    print(GetInteractionString(interaction))

$ python perceive_hints.py
bio:active-site:contact
ligand: 6 C protein: 9 N
bio:active-site:contact
ligand: 7 N protein: 9 N
bio:active-site:hbond:protein2ligand
ligand: 7 N protein: 9 N
bio:active-site:contact
ligand: 8 N protein: 9 N
bio:active-site:hbond:unpaired-protein-acceptor
protein: 3 O protein: 3 O
```

**Comment by pechersky** [comment]
And OE's visualization tools:
![image](https://github.com/user-attachments/assets/63ed89e0-55fb-4593-8dce-bad8609e4ce3)

**Comment by papillot** [comment]
So... before making any changes, I've checked with the 4.8.0 release. The geometry is correctly inferred as trigonal on the arginine guanidine Nitrogens and Carbon.
I've merged the sdf and PDB from the zip file provided by @pechersky and could assert that no HBond was perceived.
[system.pdb.zip](https://github.com/user-attachments/files/17963704/system.pdb.zip)

I've then tried with the latest master and noticed that the HBond is correctly perceived.

So, the bug has been solved recently. In fact this is due to the change made in the valence model when the hydrogens are set on a residue. 

In v4.8.0 the nitrogen on the ligand moiety is not considered as fully protonated, and the rules for N are applied which by default set 2 implicit hydrogens on this atom (!). It can't be an hydrogen acceptor and is considered as a Hbond donor instead.
<img width="617" alt="image" src="https://github.com/user-attachments/assets/f0e7395e-f009-4529-8e68-0e7d75b5def2">

In latest master, the nitrogen is considered as a HBond acceptor with 0 implicit Hs
<img width="856" alt="image" src="https://github.com/user-attachments/assets/1265e1ee-fe9a-4ba2-8c2b-4c038fd9bf90">

**Comment by pechersky** [comment]
Thanks for the debugging! Is there a branch pushed with the atom information in the labels, like in your screenshots?

I would point out that the nitrogen in the ligand should be counted as charge 0 and ideal geometry trigonal planar.

Separately, I've tried the 4.9.0 release on our data, and we've lost the H bond depiction (even with our "force ARG Ns to be conjugated). Investigating.

**Comment by papillot** [comment]
> Thanks for the debugging! Is there a branch pushed with the atom information in the labels, like in your screenshots?

Sure, it's there but you need to set the granularity to Atom/Coarse element (by clicking on the arrow cursor button in the overlayed button bar).

> I would point out that the nitrogen in the ligand should be counted as charge 0 and ideal geometry trigonal planar.

Indeed. I've investigated and the issue that I can see with the system.pdb file above is that we are entering this code:
<img width="355" alt="image" src="https://github.com/user-attachments/assets/095a976d-1f1f-45d5-a758-e7b82c304d4c">
`assignH` is false, the `valence` is 2 (because there are only 2 single bonds), which results in a `charge = -1`. This can be fixed by taking into account aromaticity at this place.

> Separately, I've tried the 4.9.0 release on our data, and we've lost the H bond depiction (even with our "force ARG Ns to be conjugated). Investigating.
I've tested the system.pdb from your example and could see the H bonds depicted on focus.
<img width="603" alt="Screenshot 2024-12-02 at 09 43 47" src="https://github.com/user-attachments/assets/fe2c5f28-f55d-448d-b2ab-89aca54fbe5e">

**Comment by pechersky** [comment]
Ah, right, I forgot a missing piece here -- we are modifying the default interaction parameters to match OE:
```
      const bondParams = {
        distanceMax: 3.2,
        donAngleDevMax: 35,
        accAngleDevMax: 30,
      };
```
And it is that smaller `donAngleDevMax` that interacts with the geometry mismatch. Since the ARG N is perceived as tetrahedral, that gives a 108 baseline angle instead of 120. So the angle here just barely misses the threshold since 108 + 35 = 143, while in the structure, it is 147.8.

**Comment by papillot** [comment]
I can't reproduce this:
- the N in ARG guanidine are flagged as **trigonal**, not tetrahedral as they should be.
- with a `donAngleDevMax` set to 35, the Hbond is perceived. I have to decrease the threshold to 32 to have it disappear.
<img width="1021" alt="Screenshot 2024-12-02 at 14 58 22" src="https://github.com/user-attachments/assets/9affbc23-d0f5-454d-bc12-e5491e3188eb">

I've investigated a bit further to understand why the 32º threshold would apply, and this is due to the closest Hydrogen atom position. There is a supplementary filter applied when Hydrogens are present, based on the assumption that the angle H-X-Acceptor is lower than the donor angle threshold. This has been added to ensure that if hydrogens are present, there is at least one in the correct direction, but in that case it might be too restrictive (again the threshold was set to 32 from the default value of 45). Maybe this angle is not necessary as long as there is an hydrogen closer to the acceptor than the donor itself.

**Comment by pechersky** [comment]
You're right, I also get the H-bond with the ARG file provided. But in the larger proprietary PDB we have, somehow the ARG is perceived as tetrahedral. Haven't been able to generate a minified example file yet.

<img width="607" alt="image" src="https://github.com/user-attachments/assets/ed315975-7200-42f4-96dd-2f692d0e5afe">

**Comment by papillot** [comment]
Just to  be sure, is `ARG` the name of the residue in the proprietary PDB or is it something else?

**Comment by pechersky** [comment]
Yes, it's ARG. I created the minimized PDB by taking the original file, deleting all non ARG lines, and resetting the resid.

**Comment by papillot** [comment]
Ok, then if it's possible to you to test a modification, could you try to make the following change in the `isConjugated` function, line 44 of mol-model-props/computed/chemistry/valence-model.ts
```ts
if (hetero && bondCount(structure, unit, index) === 4 && bondToElementCount(structure, unit, index, Elements.H) === 0) return false;
```
as a replacement for

```ts
if (hetero && bondCount(structure, unit, index) === 4) return false;
```

**Comment by papillot** [comment]
Also, are there other params which have been changed from the defaults, such as `assignH` and `assignCharge` (default is `auto`)?

**Comment by pechersky** [comment]
No, the only things I've changed are the default ideal angles, and this patch:
```
diff --git a/node_modules/molstar/lib/mol-model-props/computed/chemistry/valence-model.js b/node_modules/molstar/lib/mol-model-props/computed/chemistry/valence-model.js
index 2cbb21f..2bd2b32 100644
--- a/node_modules/molstar/lib/mol-model-props/computed/chemistry/valence-model.js
+++ b/node_modules/molstar/lib/mol-model-props/computed/chemistry/valence-model.js
@@ -8,7 +8,7 @@
 import { Unit, Bond } from '../../../mol-model/structure';
 import { isMetal } from '../../../mol-model/structure/model/properties/atomic/types';
 import { AtomGeometry, assignGeometry } from './geometry';
-import { bondCount, typeSymbol, formalCharge, bondToElementCount } from './util';
+import { bondCount, typeSymbol, formalCharge, bondToElementCount, atomId } from './util';
 import { ParamDefinition as PD } from '../../../mol-util/param-definition';
 import { isDebugMode } from '../../../mol-util/debug';
 import { SortedArray } from '../../../mol-data/int';
@@ -96,7 +96,9 @@ export function calculateHydrogensCharge(structure, unit, index, props, hasExpli
     const assignH = (props.assignH === 'always' || (props.assignH === 'auto' && !hasExplicitH && hydrogenCount === 0));
     const degree = bondCount(structure, unit, index);
     const valence = explicitValence(structure, unit, index);
-    const conjugated = isConjugated(structure, unit, index);
+    const resname = unit.model.atomicHierarchy.atoms.label_comp_id.value(unit.elements[index]);
+    const atomname = atomId(unit, index);
+    const conjugated = isConjugated(structure, unit, index) || (resname === "ARG" && (atomname === "NH1" || atomname === "NH2"));  
     const multiBond = (valence - degree > 0);
     let implicitHCount = 0;
     let geom = AtomGeometry.Unknown;
```

Trying out your suggested change

**Comment by pechersky** [comment]
The new ARG NH1:
![image](https://github.com/user-attachments/assets/650b0f79-3c61-46fe-ba87-f3c53bbc70b7)
It's the same for NH1. NE is `Charge 0`.

**Comment by papillot** [comment]
Thanks for the testing. The geometry seems to be fixed with this!

**Comment by pechersky** [comment]
Yup! Not sure what's going on with the charge aspect. But for geometry and H bond, this helps a lot, better than the ARG NH1 hard coding we were doing.

**Comment by papillot** [comment]
So, the last piece is the charge. Is it possible for you to get the values of the variables from that block for the N1 atom?
<img width="1649" alt="image" src="https://github.com/user-attachments/assets/44101986-67a2-4883-b04a-429a9576368f">
The one that seems critical is `valence`. It should be 3.

**Comment by pechersky** [comment]
I modified the function and the call site:
```
which gave the values below. `NH1` says `chg: 0`, so I'm not sure where the discrepancy in the H-bond charge calculation vs the charge-as-displayed-in-hover comes from.
```
// inside the function, before the case-switch
{
    "atomname": "NE",
    "charge": 0,
    "hydrogenCount": 1,
    "assignCharge": true,
    "assignH": false,
    "degree": 3,
    "valence": 3,
    "conjugated": false,
    "multiBond": false
}
{
    "atomname": "NH1",
    "charge": 0,
    "hydrogenCount": 2,
    "assignCharge": true,
    "assignH": false,
    "degree": 3,
    "valence": 3,
    "conjugated": false,
    "multiBond": false
}
{
    "atomname": "NH2",
    "charge": 0,
    "hydrogenCount": 2,
    "assignCharge": true,
    "assignH": false,
    "degree": 4,
    "valence": 4,
    "conjugated": false,
    "multiBond": false
}
// at the call site
{
    "atomname": "NE",
    "chg": 0,
    "implH": 0,
    "totH": 1,
    "geom": 4
}
{
    "atomname": "NH1",
    "chg": 0,
    "implH": 0,
    "totH": 2,
    "geom": 4
}
{
    "atomname": "NH2",
    "chg": 1,
    "implH": 0,
    "totH": 2,
    "geom": 4
}
```

**Comment by papillot** [comment]
Thanks for the testing!
What quizzes me is that `isConjugated` returns false for all of these... The charges calculation look alright at this place.
NH2 should have `multiBond === true`. This might be the root cause here: if there is no double bond perceived on NH2, then isConjugated will return false.

Also, NH2 has a `degree === 4` and `valence === 4`. The degree (bond count) should be 3.

I am thinking about weird stuff now: is it possible that an Hydrogen (HH21 or HH22) is somehow duplicated in the PDB file? (as they would superpose we couldn't really discern them).
